src/deepsparse/transformers/engines/nl_decoder_engine.py

```diff
@@ -0,0 +1,323 @@
+    ):
+
+        onnx_file_path, kv_cache_enabled = self.overwrite_onnx_model_inputs(
+            onnx_file_path=onnx_file_path,
+            batch_size=engine_args.get("batch_size", 1),
+            sequence_length=sequence_length,
+            input_ids_length=input_ids_length,
+        )
+
+        self.engine = create_engine(
+            onnx_file_path=onnx_file_path,
+            engine_type=engine_type,
+            engine_args=engine_args,
+            context=engine_context,
+        )
+        self.sequence_length = sequence_length
+        self.sampling_temperature = sampling_temperature
+        self.deterministic = deterministic
+        self.input_ids_length = input_ids_length
+        self.kv_cache_enabled = kv_cache_enabled
+        self.kv_cache = (
+            DecoderKVCache(use_deepsparse_cache) if kv_cache_enabled else None
+        )
+        self._num_tokens = 0  # the number of tokens processed so far
+        self._freeze_first_position = self._should_freeze_first_position(tokenizer)
+        self._session_id = generate_session_id()
+
+    @property
+    def session_id(self) -> str:
+        """
+        :return: The session id for the kv_cache if enabled
+        """
+        return self._session_id
+
+    @session_id.setter
+    def session_id(self, session_id: str):
+        """
+        :param session_id: The session id to set for the kv_cache
+        """
+        self._session_id = session_id
+
+    @property
+    def onnx_input_names_no_cache(self) -> List[str]:
+        """
+        :return: The input names for the onnx model, excluding
+            the potential kv cache inputs
+        """
+        return [
+            name
+            for name in self.engine.input_names
+            if not name.startswith(_CACHE_INPUT_NAME)
+        ]
+
+    def __call__(
+        self,
+        inp: List[numpy.ndarray],
+        val_inp: bool = True,
+    ) -> Tuple[numpy.ndarray, numpy.ndarray]:
+        """
+        The main entry point for running the engine.
+
+        :param inp: The input to run the engine with. We expect a
+            list of numpy arrays that contain the input ids,
+            attention mask, and position ids (optionally)
+        :param val_inp: Whether the input is for validation or not
+        :return: The generated token and corresponding logits
+        """
+        if self.kv_cache:
+            # if kv cache is enabled, we need to add the kv cache state
+            # to the input
+            inp = self._add_kv_cache_to_input(inp)
+
+        out = self.engine.run(inp, val_inp)
+
+        self._num_tokens += self.input_ids_length
+
+        if self.kv_cache:
+            logits, *kv_cache_state = out
+            self._update_kv_cache(
+                kv_cache_state=kv_cache_state,
+            )
+        else:
+            logits = out[0]
+            B, S, V = logits.shape  # batch, sequence, vocab
+            logits = logits[:, -1, :].reshape(B, 1, V)  # only take the last token
```

Comment:
why do we slice logits here but not in kv_cache mode? if it's because we expect different upstream usage, shouldn't we let the upstream pipeline handle that?

Reply:
Good one did not catch this, since never tested the kv cache run with input_ids!= 1